internal/encryption/encryption.go

```diff
@@ -0,0 +1,364 @@
+	for name, kps := range enc.keyValues {
+		kpMap[name] = cty.ObjectVal(kps)
+	}
+	enc.ctx.Variables["key_provider"] = cty.ObjectVal(kpMap)
+
+	return diags
+}
+
+func (enc *encryption) setupMethods() hcl.Diagnostics {
+	var diags hcl.Diagnostics
+	for _, m := range enc.cfg.Methods {
+		diags = append(diags, enc.setupMethod(m)...)
+	}
+
+	// Regen ctx
+	mMap := make(map[string]cty.Value)
+	for name, ms := range enc.methodValues {
+		mMap[name] = cty.ObjectVal(ms)
+	}
+	enc.ctx.Variables["method"] = cty.ObjectVal(mMap)
+
+	return diags
+}
+
+func (enc *encryption) setupMethod(cfg MethodConfig) hcl.Diagnostics {
+	// Ensure cfg.Type is in methodValues
+	if _, ok := enc.methodValues[cfg.Type]; !ok {
+		enc.methodValues[cfg.Type] = make(map[string]cty.Value)
+	}
+
+	// Lookup definition
+	encryptionMethod, err := enc.reg.GetMethod(method.ID(cfg.Type))
+	if err != nil {
+		return hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Unknown method type",
+			Detail:   err.Error(),
+		}}
+	}
+
+	methodcfg := encryptionMethod.ConfigStruct()
+
+	// TODO we could use varhcl here to provider better error messages
+	diags := gohcl.DecodeBody(cfg.Body, enc.ctx, methodcfg)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// Map from EvalContext vars -> Descriptor
+	mIdent := fmt.Sprintf("method.%s.%s", cfg.Type, cfg.Name)
+	enc.methodValues[cfg.Type][cfg.Name] = cty.StringVal(mIdent)
+	enc.methods[mIdent], err = methodcfg.Build()
+	if err != nil {
+		// TODO this error handling could use some work
+		return hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Method configuration failed",
+			Detail:   err.Error(),
+		}}
+	}
+
+	return diags
+
+}
+
+func (enc *encryption) setupTargets() hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	if enc.cfg.StateFile != nil {
+		m, mDiags := enc.setupTarget(enc.cfg.StateFile, "statefile")
+		diags = append(diags, mDiags...)
+		enc.stateFile = NewState(m)
+	}
+	if enc.cfg.PlanFile != nil {
+		m, mDiags := enc.setupTarget(enc.cfg.PlanFile, "planfile")
+		diags = append(diags, mDiags...)
+		enc.planFile = NewPlan(m)
+	}
+	if enc.cfg.Backend != nil {
+		m, mDiags := enc.setupTarget(enc.cfg.Backend, "backend")
+		diags = append(diags, mDiags...)
+		enc.backend = NewState(m)
+	}
+
+	if enc.cfg.Remote != nil {
+		if enc.cfg.Remote.Default != nil {
+			m, mDiags := enc.setupTarget(enc.cfg.Remote.Default, "remote_data_source.default")
+			diags = append(diags, mDiags...)
+			enc.remoteDefault = NewState(m)
+		}
+		for _, target := range enc.cfg.Remote.Targets {
+			m, mDiags := enc.setupTarget(&TargetConfig{
+				Enforced: target.Enforced,
+				Method:   target.Method,
+				Fallback: target.Fallback,
+			}, "remote_data_source."+target.Name)
+			diags = append(diags, mDiags...)
+			enc.remote[target.Name] = NewState(m)
+		}
+	}
+
+	return diags
+}
+
+func (enc *encryption) setupTarget(cfg *TargetConfig, name string) ([]method.Method, hcl.Diagnostics) {
```

Comment:
I am a bit confused as to why `setupTarget` returns an array of `Method`s. Seems like it'll be a 1-2 sized array, the first being the actual encryption method, and the optional second being the fallback method. 

If that's the case then the type here and in other usages (In `NewPlan`, etc.) is a bit confusing, especially since we've explicitly declared we don't want more than 1 fallback

If I understand correctly, then this means the fallback logic is not yet implemented (at least for plan encryption), correct?